asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py

```diff
@@ -0,0 +1,273 @@
+
+    if target_counts:
+        orig_counts_list = epc_subset[feature_name].value_counts().to_dict()
+        orig_counts = dict(
+            zip(orig_counts_list[feature_name], orig_counts_list["count"])
+        )
+        orig_proportions = calculate_proportions(orig_counts)
+        error_metrics_orig = get_error_metrics(
+            orig_counts, orig_proportions, target_counts, target_proportions
+        )
+        if not all(epc_subset["weight"].is_null()):
+            # Get the sum of the weights for each feature category (not technically a 'count')
+            reweighted_counts_list = (
+                epc_subset.group_by(feature_name).agg(pl.col("weight").sum()).to_dict()
+            )
+            reweighted_counts = dict(
+                zip(
+                    reweighted_counts_list[feature_name],
+                    reweighted_counts_list["weight"],
+                )
+            )
+            reweighted_proportions = calculate_proportions(reweighted_counts)
+            error_metrics_reweighted = get_error_metrics(
+                reweighted_counts,
+                reweighted_proportions,
+                target_counts,
+                target_proportions,
+            )
+
+            average_error_reduction = get_error_reduction(
+                orig_proportions, reweighted_proportions, target_proportions
+            )
+
+            prop_reweighted_counts_list = (
+                epc_subset.group_by(feature_name)
+                .agg(pl.col("proportional_weight").sum())
+                .to_dict()
+            )
+            prop_reweighted_counts = dict(
+                zip(
+                    prop_reweighted_counts_list[feature_name],
+                    prop_reweighted_counts_list["proportional_weight"],
+                )
+            )
+            prop_reweighted_proportions = calculate_proportions(prop_reweighted_counts)
+            error_metrics_prop_reweighted = get_error_metrics(
+                prop_reweighted_counts,
+                prop_reweighted_proportions,
+                target_counts,
+                target_proportions,
+            )
+
+        else:
+            # The weights may have been all Null
+            error_metrics_reweighted = {}
+            error_metrics_prop_reweighted = {}
+            average_error_reduction = None
+    else:
+        error_metrics_orig = {}
+        error_metrics_reweighted = {}
+        error_metrics_prop_reweighted = {}
+        average_error_reduction = None
+
+    final_metrics = {}
+    for metric in main_error_metrics:
+        final_metrics[metric] = {
+            "unweight": error_metrics_orig.get(metric, None),
+            "reweight": error_metrics_reweighted.get(metric, None),
+            "reweight-prop": error_metrics_prop_reweighted.get(metric, None),
+            "error_reduction": average_error_reduction,
+        }
+    return final_metrics
+
+
+def load_epc_reweights(
```

Comment:
You can put the cleaning function in `pipeline/enhance_epc/prepare_epc.py` :)

Reply:
good idea!